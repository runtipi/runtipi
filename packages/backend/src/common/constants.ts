export const APP_DIR = '/app';
export const DATA_DIR = '/data';
export const APP_DATA_DIR = '/app-data';

export const SESSION_COOKIE_NAME = 'tipi.sid';
export const SESSION_COOKIE_MAX_AGE = 1000 * 60 * 60 * 24;

export const ARCHITECTURES = ['arm64', 'amd64'] as const;
export type Architecture = (typeof ARCHITECTURES)[number];

export const LATEST_RELEASE_URL = 'https://api.github.com/repos/runtipi/runtipi/releases/latest';
