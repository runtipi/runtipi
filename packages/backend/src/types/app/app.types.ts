export type AppUrn = string & { __type: 'urn' };
