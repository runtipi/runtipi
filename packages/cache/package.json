{
  "name": "@runtipi/cache",
  "version": "1.0.0",
  "description": "",
  "main": "./src/index.ts",
  "module": "./src/index.ts",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 0"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "@runtipi/shared": "workspace:^",
    "ioredis": "^5.4.1"
  },
  "devDependencies": {
    "vitest": "^2.1.1"
  }
}
