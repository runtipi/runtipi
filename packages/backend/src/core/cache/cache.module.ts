import { Global, Module } from '@nestjs/common';
import { CacheService } from './cache.service.js';

@Global()
@Module({
  imports: [],
  controllers: [],
  providers: [CacheService],
  exports: [CacheService],
})
export class CacheModule {}
