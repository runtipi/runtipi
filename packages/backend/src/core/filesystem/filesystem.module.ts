import { Global, Module } from '@nestjs/common';
import { FilesystemService } from './filesystem.service.js';

@Global()
@Module({
  imports: [],
  providers: [FilesystemService],
  exports: [FilesystemService],
})
export class FilesystemModule {}
