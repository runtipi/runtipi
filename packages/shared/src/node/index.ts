import { execAsync } from './helpers/exec-async';
import { pathExists } from './helpers/fs-helpers';

import { Logger } from './logger/FileLogger';
import type { ILogger } from './logger/Logger.interface';
import { AppDataService, type IAppDataService } from './modules/app/app-data-service';
import { AppFileAccessor, type IAppFileAccessor } from './modules/app/app-file-accessor';
import { BackupManager, type IBackupManager } from './modules/backup/backup-manager';

export {
  execAsync,
  pathExists,
  Logger,
  AppDataService,
  type ILogger,
  type IAppDataService,
  AppFileAccessor,
  type IAppFileAccessor,
  BackupManager,
  type IBackupManager,
};
