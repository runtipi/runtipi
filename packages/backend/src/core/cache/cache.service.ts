import { DatabaseSync } from 'node:sqlite';
import { Injectable } from '@nestjs/common';

export const ONE_DAY_IN_SECONDS = 60 * 60 * 24;

@Injectable()
export class CacheService {
  private db: DatabaseSync;

  constructor() {
    this.db = new DatabaseSync('/cache/cache.sqlite');

    const tableCheck = this.db.prepare("SELECT name FROM sqlite_master WHERE type='table'").get();

    if (!tableCheck) {
      this.db.exec('CREATE TABLE keyv (key TEXT PRIMARY KEY, value TEXT)');
    }
  }

  public set(key: string, value: string, expiration = ONE_DAY_IN_SECONDS) {
    const stmt = this.db.prepare('INSERT OR REPLACE INTO keyv (key, value) VALUES (?, ?)');
    stmt.run(key, JSON.stringify({ value, expiration: Math.max(Date.now() + expiration * 1000, expiration) }));
  }

  public get(key: string) {
    const query = this.db.prepare('SELECT * FROM keyv WHERE key = ?');
    const row = query.get(key) as { value: string } | undefined;

    if (!row) {
      return undefined;
    }

    const { value, expiration = 0 } = JSON.parse(row.value) as { value: string; expiration: number };
    if (expiration < Date.now()) {
      this.del(key);
      return undefined;
    }

    return value;
  }

  public del(key: string) {
    const stmt = this.db.prepare('DELETE FROM keyv WHERE key = ?');
    stmt.run(key);
  }

  public async getByPrefix(prefix: string) {
    try {
      const query = this.db.prepare('SELECT * FROM keyv WHERE key LIKE ?');
      const rows = query.all(`cache:${prefix}%`) as { key: string; value: string }[];

      return rows.map((row) => ({ key: row.key, val: JSON.parse(row.value).value }));
    } catch (error) {
      console.error(error);
      return [];
    }
  }

  public clear() {
    const stmt = this.db.prepare('DELETE FROM keyv');
    stmt.run();
  }
}
