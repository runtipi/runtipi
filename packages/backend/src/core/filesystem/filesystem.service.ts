import fs from 'node:fs';
import { EOL } from 'node:os';
import path from 'node:path';
import { APP_DATA_DIR, APP_DIR, DATA_DIR } from '@/common/constants';
import { LoggerService } from '@/core/logger/logger.service';
import { Injectable } from '@nestjs/common';
import { ZodSchema } from 'zod';

@Injectable()
export class FilesystemService {
  constructor(private readonly logger: LoggerService) {}

  private getSafeFilePath(filePath: string): string {
    // Define allowed directories as absolute paths
    const allowedDirs = [
      path.resolve(APP_DIR),
      path.resolve(APP_DATA_DIR),
      path.resolve(DATA_DIR),
      path.resolve('/host/proc/'),
      path.resolve('/tmp/'),
    ];

    // Resolve and normalize the file path to an absolute path
    const resolvedPath = path.resolve(filePath);

    for (const dir of allowedDirs) {
      if (path.relative(dir, resolvedPath).startsWith('..')) {
        continue; // If relative path starts with '..', it's outside the allowed dir
      }

      return resolvedPath;
    }

    this.logger.error(`File path "${filePath}" is not allowed. Resolved: "${resolvedPath}"`);
    throw new Error('File path is not allowed');
  }

  async readJsonFile<T>(filePath: string, schema?: ZodSchema<T>): Promise<T | null> {
    try {
      const fileContent = await fs.promises.readFile(this.getSafeFilePath(filePath), 'utf8');
      const parsedContent = JSON.parse(fileContent);

      if (schema) {
        const validatedContent = schema.safeParse(parsedContent);
        if (!validatedContent.success) {
          this.logger.debug(`File ${this.getSafeFilePath(filePath)} validation error:`, validatedContent.error);
          return null;
        }
        return validatedContent.data;
      }

      return parsedContent;
    } catch (error) {
      this.logger.error(`Error reading file ${this.getSafeFilePath(filePath)}: ${error}`);
      return null;
    }
  }

  async readTextFile(filePath: string): Promise<string | null> {
    try {
      return await fs.promises.readFile(this.getSafeFilePath(filePath), 'utf8');
    } catch (error) {
      this.logger.error(`Error reading file ${this.getSafeFilePath(filePath)}: ${error}`);
      return null;
    }
  }

  async readBinaryFile(filePath: string): Promise<Buffer | null> {
    try {
      return await fs.promises.readFile(this.getSafeFilePath(filePath));
    } catch (error) {
      this.logger.error(`Error reading file ${this.getSafeFilePath(filePath)}: ${error}`);
      return null;
    }
  }

  async writeJsonFile<T>(filePath: string, data: T): Promise<boolean> {
    try {
      await fs.promises.writeFile(this.getSafeFilePath(filePath), `${JSON.stringify(data, null, 2)}${EOL}`, 'utf8');
      return true;
    } catch (error) {
      this.logger.error(`Error writing file ${this.getSafeFilePath(filePath)}: ${error}`);
      return false;
    }
  }

  async writeTextFile(filePath: string, content: string): Promise<boolean> {
    try {
      await fs.promises.mkdir(this.getSafeFilePath(filePath.split('/').slice(0, -1).join('/')), { recursive: true });
      await fs.promises.writeFile(this.getSafeFilePath(filePath), `${content}${EOL}`, 'utf8');
      return true;
    } catch (error) {
      this.logger.error(`Error writing file ${this.getSafeFilePath(filePath)}: ${error}`);
      return false;
    }
  }

  async pathExists(filePath: string): Promise<boolean> {
    return fs.promises
      .access(this.getSafeFilePath(filePath))
      .then(() => true)
      .catch(() => false);
  }

  async copyFile(src: string, dest: string): Promise<boolean> {
    try {
      await fs.promises.copyFile(this.getSafeFilePath(src), this.getSafeFilePath(dest));
      return true;
    } catch (error) {
      this.logger.error(`Error copying file from ${this.getSafeFilePath(src)} to ${this.getSafeFilePath(dest)}: ${error}`);
      return false;
    }
  }

  async createDirectory(dirPath: string): Promise<boolean> {
    try {
      await fs.promises.mkdir(this.getSafeFilePath(dirPath), { recursive: true });
      return true;
    } catch (error) {
      this.logger.error(`Error creating directory ${this.getSafeFilePath(dirPath)}: ${error}`);
      return false;
    }
  }

  async createDirectories(dirPaths: string[]): Promise<boolean> {
    for (const dirPath of dirPaths) {
      if (!(await this.createDirectory(this.getSafeFilePath(dirPath)))) {
        return false;
      }
    }
    return true;
  }

  async copyDirectory(src: string, dest: string, options: fs.CopyOptions = {}): Promise<boolean> {
    try {
      await fs.promises.cp(this.getSafeFilePath(src), this.getSafeFilePath(dest), { recursive: true, ...options });
      return true;
    } catch (error) {
      this.logger.error(`Error copying directory from ${this.getSafeFilePath(src)} to ${this.getSafeFilePath(dest)}: ${error}`);
      return false;
    }
  }

  async removeDirectory(dirPath: string): Promise<boolean> {
    try {
      await fs.promises.rm(this.getSafeFilePath(dirPath), { recursive: true, force: true, maxRetries: 3, retryDelay: 100 });
      return true;
    } catch (error) {
      this.logger.error(`Error removing directory ${this.getSafeFilePath(dirPath)}: ${error}`);
      return false;
    }
  }

  async removeFile(filePath: string): Promise<boolean> {
    try {
      await fs.promises.unlink(this.getSafeFilePath(filePath));
      return true;
    } catch (error) {
      this.logger.error(`Error removing file ${this.getSafeFilePath(filePath)}: ${error}`);
      return false;
    }
  }

  async listFiles(dirPath: string): Promise<string[]> {
    try {
      return await fs.promises.readdir(this.getSafeFilePath(dirPath));
    } catch (error) {
      this.logger.error(`Error listing files in ${this.getSafeFilePath(dirPath)}: ${error}`);
      return [];
    }
  }

  async isDirectory(dirPath: string): Promise<boolean> {
    return (await fs.promises.lstat(this.getSafeFilePath(dirPath))).isDirectory();
  }

  async createTempDirectory(prefix: string): Promise<string | null> {
    return fs.promises.mkdtemp(prefix);
  }

  async getStats(filePath: string) {
    return await fs.promises.stat(this.getSafeFilePath(filePath));
  }
}
