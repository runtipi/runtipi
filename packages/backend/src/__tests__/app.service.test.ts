import fs from 'node:fs';
import { AppService } from '@/app.service';
import { APP_DATA_DIR, APP_DIR, DATA_DIR, LATEST_RELEASE_URL } from '@/common/constants';
import { CacheService } from '@/core/cache/cache.service';
import { ConfigurationService } from '@/core/config/configuration.service';
import { FilesystemService } from '@/core/filesystem/filesystem.service';
import type { FsMock } from '@/tests/__mocks__/fs';
import { faker } from '@faker-js/faker';
import { Test } from '@nestjs/testing';
import { fromPartial } from '@total-typescript/shoehorn';
import { http, HttpResponse } from 'msw';
import { setupServer } from 'msw/node';
import { beforeAll, beforeEach, describe, expect, it } from 'vitest';
import { mock } from 'vitest-mock-extended';

const server = setupServer();

describe('AppService', () => {
  let appService: AppService;
  let configurationService = mock<ConfigurationService>();
  let cacheService = mock<CacheService>();

  beforeAll(() => {
    server.listen();
    server.use(
      http.get(LATEST_RELEASE_URL, () => {
        return HttpResponse.json({
          tag_name: 'latest',
          body: 'body',
        });
      }),
    );
  });

  beforeEach(async () => {
    const moduleRef = await Test.createTestingModule({
      providers: [AppService, FilesystemService],
    })
      .useMocker(mock)
      .compile();

    appService = moduleRef.get(AppService);
    configurationService = moduleRef.get(ConfigurationService);
    cacheService = moduleRef.get(CacheService);
  });

  describe('getVersion', () => {
    it('should return the version', async () => {
      // arrange
      const version = faker.system.semver();
      configurationService.getConfig.mockReturnValueOnce(fromPartial({ version }));

      // act
      const result = await appService.getVersion();

      // assert
      expect(result.current).toBe(version);
    });

    it('shoult return version from cache if set', async () => {
      // arrange
      const version = faker.system.semver();
      const latest = faker.system.semver();
      configurationService.getConfig.mockReturnValueOnce(fromPartial({ version }));
      cacheService.get.calledWith('latestVersion').mockResolvedValueOnce(latest);
      cacheService.get.calledWith('latestVersionBody').mockResolvedValueOnce('body');

      // act
      const result = await appService.getVersion();

      // assert
      expect(result.current).toBe(version);
      expect(result.latest).toBe(latest);
      expect(result.body).toBe('body');
    });

    it('should fetch latest version from github if not in cache', async () => {
      // arrange
      const version = faker.system.semver();
      const latest = faker.system.semver();
      const body = faker.lorem.paragraph();
      configurationService.getConfig.mockReturnValueOnce(fromPartial({ version }));
      cacheService.get.calledWith('latestVersion').mockResolvedValueOnce(undefined);
      cacheService.get.calledWith('latestVersionBody').mockResolvedValueOnce(undefined);

      server.use(
        http.get(LATEST_RELEASE_URL, () => {
          return HttpResponse.json({
            tag_name: latest,
            body,
          });
        }),
      );

      // act
      const result = await appService.getVersion();

      // assert
      expect(result.current).toBe(version);
      expect(result.latest).toBe(latest);
      expect(result.body).toBe(body);
    });

    it('should return current version if cache fails', async () => {
      // arrange
      const version = faker.system.semver();
      configurationService.getConfig.mockReturnValueOnce(fromPartial({ version }));
      cacheService.get.calledWith('latestVersion').mockRejectedValueOnce(new Error('error'));

      // act
      const result = await appService.getVersion();

      // assert
      expect(result.current).toBe(version);
      expect(result.latest).toBe(version);
      expect(result.body).toBe('');
    });

    it('should return current version if fetch fails', async () => {
      // arrange
      const version = faker.system.semver();
      configurationService.getConfig.mockReturnValueOnce(fromPartial({ version }));
      cacheService.get.calledWith('latestVersion').mockResolvedValueOnce(undefined);

      server.use(
        http.get(LATEST_RELEASE_URL, () => {
          return new HttpResponse('error', { status: 500 });
        }),
      );

      // act
      const result = await appService.getVersion();

      // assert
      expect(result.current).toBe(version);
      expect(result.latest).toBe(version);
      expect(result.body).toBe('');
    });
  });

  describe('copyAssets', () => {
    it('should create base folder structure', async () => {
      // arrange
      const appDir = APP_DIR;
      const dataDir = DATA_DIR;
      const appDataDir = APP_DATA_DIR;
      const directories = { appDir, dataDir, appDataDir };
      configurationService.getConfig.mockReturnValueOnce(fromPartial({ directories, userSettings: { persistTraefikConfig: false } }));

      // act
      await appService.copyAssets();

      // assert
      expect((fs as unknown as FsMock).tree()).toMatchSnapshot();
    });
  });
});
