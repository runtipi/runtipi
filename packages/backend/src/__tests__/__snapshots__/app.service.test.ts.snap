// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppService > copyAssets > should create base folder structure 1`] = `
"/
├─ app-data/
└─ data/
   ├─ .env
   ├─ apps/
   ├─ backups/
   ├─ media/
   │  ├─ data/
   │  │  ├─ books/
   │  │  ├─ comics/
   │  │  ├─ images/
   │  │  ├─ movies/
   │  │  ├─ music/
   │  │  ├─ podcasts/
   │  │  ├─ roms/
   │  │  └─ tv/
   │  ├─ downloads/
   │  │  ├─ complete/
   │  │  ├─ incomplete/
   │  │  └─ watch/
   │  ├─ torrents/
   │  │  ├─ complete/
   │  │  ├─ incomplete/
   │  │  └─ watch/
   │  └─ usenet/
   │     ├─ complete/
   │     ├─ incomplete/
   │     └─ watch/
   ├─ repos/
   ├─ state/
   │  └─ seed
   └─ traefik/
      ├─ dynamic/
      ├─ shared/
      └─ tls/"
`;
